lib/core/include/metadata.hpp

```diff
@@ -0,0 +1,232 @@
+    #define rxModAVUMetadata rcModAVUMetadata
+#endif
+
+#include "filesystem.hpp"
+#include "fmt/format.h"
+
+namespace irods::experimental::metadata {
+
+    struct avu {
+        std::string attribute;
+        std::string value;
+        std::string units;
+    };
+
+    auto operator==(const avu& _lhs, const avu& _rhs) {
+        bool eq = _lhs.attribute == _rhs.attribute &&
+                  _lhs.value     == _rhs.value &&
+                  _lhs.units     == _rhs.units;
+        return eq;
+    }
+
+    namespace tokens {
+        const std::string none{};
+        const std::string collection{"-C"};
+        const std::string object{"-d"};
+        const std::string user{"-u"};
+        const std::string resource{"-R"};
+    }
+
+    namespace {
+
+        using bf = boost::format;
+
+        using entity_type = irods::experimental::entity::type;
+
+        struct EnumClassHash
+        {
+            template <typename T>
+                std::size_t operator()(T t) const
+                {
+                    return static_cast<std::size_t>(t);
+                }
+        };
+
+        const std::unordered_map<entity_type, std::string, EnumClassHash> type_to_token {
+            {entity_type::collection, tokens::collection},
+            {entity_type::object,     tokens::object},
+            {entity_type::user,       tokens::user},
+            {entity_type::resource,   tokens::resource}
+        };
+
+        const std::unordered_map<std::string, entity_type> token_to_type {
+            {tokens::collection, entity_type::collection},
+            {tokens::object,     entity_type::object},
+            {tokens::user,       entity_type::user},
+            {tokens::resource,   entity_type::resource}
+        };
+
+        auto mod_avu_meta(rxComm_t& comm, const avu& _md, const avu& _dmd, entity_type _et, const std::string& _tgt, const std::string& _op)
+        {
+
+            modAVUMetadataInp_t inp{
+                const_cast<char*>(_op.c_str())
+              , const_cast<char*>(type_to_token.at(_et).c_str())
+              , const_cast<char*>(_tgt.c_str())
+              , const_cast<char*>(_md.attribute.c_str())
+              , const_cast<char*>(_md.value.c_str())
+              , const_cast<char*>(_md.units.c_str())
+              , const_cast<char*>(_dmd.attribute.c_str())
+              , const_cast<char*>(_dmd.value.c_str())
+              , const_cast<char*>(_dmd.units.c_str())
+              , nullptr // unused arg9
+            };
+
+            return rxModAVUMetadata(&comm, &inp);
+
+        } // mod_avu_meta
+
+    } // namespace
+
+
+    auto to_type(const std::string& _t)
+    {
+
+        if(token_to_type.find(_t) == token_to_type.end()) {
+            return entity_type::none;
+        }
+
+        return token_to_type.at(_t);
+
+    } // to_type
+
+
+    auto to_token(entity_type _t)
+    {
+
+        if(type_to_token.find(_t) == type_to_token.end()) {
+            return tokens::none;
+        }
```

Comment:
this is unnecessary given that the incoming enumeration guarantees that a valid type exists.